doc/dev/msgr2.rst

```diff
@@ -0,0 +1,199 @@
+  consists of a tag (type code), payload, and (if signing or
+  encryption is enabled) some other fields.  See below for the
+  structure.
+* *tag*: a single-byte type code associated with a frame.  The tag
+  determines the structure of the payload.
+
+Phases
+------
+
+A connection has four distinct phases:
+
+#. banner
+#. authentication
+#. message flow handshake
+#. message exchange
+
+Banner
+------
+
+Both the client and server, upon connecting, send a banner::
+
+  "ceph %x %x\n", protocol_features_suppored, protocol_features_required
+
+The protocol features are a new, distinct namespace.  Initially no
+features are defined or required, so this will be "ceph 0 0\n".
+
+If the remote party advertises features we don't support, we can
+disconnect.
+
+Authentication
+--------------
+
+A series of frames between client and server of the form::
+
+  tag byte (TAG_*)
+  payload
+
+where the payload is determiend by the tag.
+
+* TAG_AUTH_METHODS (server only): list authentication methods (none, cephx, ...)::
+
+    __le32 num_methods;
+    __le32 methods[num_methods];   // CEPH_AUTH_{NONE, CEPHX}
+
+* TAG_AUTH_SET_METHOD (client only): set auth method for this connection::
+
+    __le32 method;
+
+  - The selected auth method determines the sig_size and block_size in any
+    subsequent messages (TAG_AUTH_DONE and non-auth messages).
+
+* TAG_AUTH_BAD_METHOD (server only): reject client-selected auth method::
+
+    __le32 method
+
+* TAG_AUTH: client->server or server->client auth message::
+
+    __le32 len;
+    method specific payload
+
+* TAG_AUTH_DONE::
+    
+    __le64 flags
+      FLAG_ENCRYPTED  1
+      FLAG_SIGNED     2
+    signature
+
+  - The client first says AUTH_DONE, and the server replies to
+    acknowledge it.
+
+
+Message frame format
+--------------------
+
+Each frame can take one of two forms.  If FLAG_SIGNED or
+FLAG_ENCRYPTED has been specified::
+
+  confounder (block_size bytes of random garbage)
+  __le32 length
+  tag byte
+  payload
+  signature (sig_size bytes)
+  more confounder padding (to pad data from start of __le32 length out to block size)
+
+Note that the padding ensures that the total frame (with or without
+the leading confounder) is a multiple of the auth method's block_size.
+This is usually something like 16 bytes.
+
+If neither FLAG_SIGNED or FLAG_ENCRYPTED is specified, things are simple::
+
+  tag byte
+  payload
+    
+Message flow handshake
+----------------------
+
+In this phase the peers identify each other and (if desired) reconnect to
+an established session.
+
+* TAG_IDENT: identify ourselves::
+
+    entity_addrvec_t addr(s)
+    __u8   my type (CEPH_ENTITY_TYPE_*)
+    __le32 protocol version
+    __le64 features supported (CEPH_FEATURE_* bitmask)
+    __le64 features required (CEPH_FEATURE_* bitmask)
+    __le64 flags (CEPH_MSG_CONNECT_* bitmask)
+    __le64 cookie (a client identifier, assigned by the sender. unique on the sender.)
```

Comment:
We spent a lot of time talking about this and I'm not sure if we actually came to consensus.

AFAIK, the only reason to have this is to allow the client to quickly identify itself to the server when re-establishing an existing session. So on first connect, it might as well be zeroed out.
But when reconnecting, it needs to be something that is 1) a unique identifier to the server, 2) has been authenticated in both directions.

Which makes me think that:
*) on first connect, the client fills in zero.
*) On any connect, the server replies to the client with both the client-supplied cookie AND a new cookie for the client to use in future reconnects.
(And if the server doesn't fill in a new one, then the client just doesn't get to do a fast reconnect)

Alternatively we could have a different tag with a different struct for reconnects. Or the server could not include the client-supplied cooke in its authenticated reply, but I'm worried (perhaps baselessly) about replay attacks in that case.